etl/grapher_import.py

```diff
@@ -575,6 +575,19 @@ def cleanup_ghost_variables(engine: Engine, dataset_id: int, upserted_variable_i
         return True
 
 
+def _raise_error_for_deleted_variables(rows: pd.DataFrame) -> bool:
+    """If we run into ghost variables that are still used in charts, should we raise an error?"""
+    # raise an error if on staging server
+    if config.ENV == "staging":
+        # It's possible that we merged changes to ETL, but the staging server still uses old charts. In
+        # that case, we first check that the charts were really modified on our staging server.
+        modified_charts = ChartDiffsLoader(config.OWID_ENV.get_engine(), production_or_master_engine()).df
+        return bool(set(modified_charts.index) & set(rows.chartId))
```

Comment:
I tried following the logic here, but didn't really come to an understanding. Could you help me understand it?

So, my logic is:

1) Before we had something like: "raise error if there are ghost variables (i.e. `rows`)"

2) No, you are proposing to evaluate `bool(set(modified_charts.index))` in addition. This should return `True` whenever `modified_charts` is not empty, right? (In that case wouldn't it be enough to just use `not modified_charts.empty`?). 

3) Still, if `modified_charts` is not empty, why is that a potential sign to raise an error? Isn't `modified_charts` containing a row per each chart that has been modified (regardless of it being approved, etc.)?

I think I'm missing something here. Thanks!

Reply:
> No, you are proposing to evaluate bool(set(modified_charts.index)) in addition. This should return True whenever modified_charts is not empty, right? (In that case wouldn't it be enough to just use not modified_charts.empty?).

`set(modified_charts.index)` contains chart ids that have been modified on staging (either via config, data or metadata) and `set(rows.chartId)` contains problematic chart ids. Previously, we raised an error if `set(rows.chartId)` wasn't empty, but the problem was that sometimes it was empty because of race conditions between staging & prod.

`bool(set(modified_charts.index) & set(rows.chartId))` returns True if there are any problematic charts AND those charts have been also modified on the staging server.

I could have also written it as:
```
if rows and any(row.chartId in modified_charts.index):
  return True
```

Perhaps it's me who's missing something and I misunderstood your question.

Reply:
oh my god. I miss-read one of the parents. I read `bool(set(modified_charts.index)) & set(rows.chartId)`.

NVM, I got it, makes total sense.